Add convenience method to use LangChain community tools

matthewfranglen opened the pull request:
References #1831 

This does work however the style and approach are very much open for discussion. In particular:

* this creates a function just so it can be inspected, as part of that it has a very bad mapping from json schema types back to python.
* the `pydantic_ai.tool.Tool` class has no base class that describes the desired behaviour.
* the LangChain BaseTool defines json schemas for the args and similar which could more directly be translated to the `pydantic_ai.tool.Tool` class.
* future integrations of this type might benefit from a similar approach.

To make my PR simple I have introduced a minimal change. I think it would be good to discuss a more maintainable and extensible approach.

DouweM commented:
@matthewfranglen Thanks Matthew! As you've noted, the most brittle part of the implementation is mapping back from the JSON schema we get from LangChain to a function that can be turned into a JSON schema _and a Pydantic validator_ by PydanticAI.

In https://github.com/pydantic/pydantic-ai/pull/1785, we've got a pending refactor of `Tool` to take a `FunctionSchema` object with `json_schema`, `description`, and `validator` properties and a `call` method. 

The cleanest implementation would likely be to subclass `FunctionSchema` and override the `json_schema`, and then build a `validator` from that schema, circumventing all of the function-signature magic. 

I'm not sure if there's a way to build a `SchemaValidator` from a JSON schema. @Viicos @Kludex Is there?

matthewfranglen commented:
@DouweM thanks for the feedback. I'll review the linked PR tomorrow and see if I can refit this to use the `FunctionSchema`.

Viicos commented:
> I'm not sure if there's a way to build a `SchemaValidator` from a JSON schema. @Viicos @Kludex Is there?

We discussed about it a while ago. I started on a simple implementation but it is a lot of work to get it working, and many JSON Schema concepts can't be expressed using core schemas so it is hard to not end up with a half-working solution.

DouweM commented:
@Viicos OK, I think it'll be worth adding that functionality.

@matthewfranglen To keep making progress on this PR, for now I think we can get away without validation and trust that the model outputs JSON matching the schema (especially if using OpenAIModel with strict mode).

matthewfranglen commented:
@DouweM I'll be able to start working on this again in a few hours. As I understand it the requirement is to refit this to use the FunctionSchema, except for now without a validator.

DouweM commented:
@matthewfranglen That's right, and I'd like to have a generic way to build a tool from a JSON schema and a some function to be called (which may not have the same arguments as the JSON schema -- it could be `*args, **kwargs` if they get forwarded somewhere), _and_ a LangChain-specific method built on top of that.

matthewfranglen commented:
@DouweM I've done my best with this then.

* There is now `Tool.from_function` which takes a function and a json schema and constructs a tool from that, using the name and docstring of the function. This uses your FunctionSchema and has a permissive validator. To get this I merged your in progress branch so this PR now depends on that.
* The `Tool.from_langchain` uses the above
* There are tests for the `from_function` code

I've added tests for the code paths that are not covered when using the Agent.

matthewfranglen commented:
Thinking about it, `from_function` should have a different name. Most of the tool examples are of the form `Tool(function)` so the name doesn't really describe what is different about it. Thoughts?

DouweM commented:
@matthewfranglen What about `Tool.from_schema` as the main difference is that you pass in your own schema.

Also my PR has now been merged so please merge main again!

matthewfranglen commented on pydantic_ai_slim/pydantic_ai/tools.py:
```diff
@@ -289,28 +306,97 @@ async def prep_my_tool(
+            var_positional_field=None,
+        )
+
+        return Tool(
```
Self was introduced in 3.11 so 3.9 and 3.10 will fail the tests.

matthewfranglen commented:
@Viicos I've addressed your feedback, thank you for providing it.

@DouweM I think that name makes a lot of sense, I have updated it. I have also merged main.

Is this ready to merge?

DouweM commented on pydantic_ai_slim/pydantic_ai/tools.py:
```diff
@@ -9,16 +10,34 @@
+        # {'dir_path': {'default': '.', 'description': 'Subdirectory to search in.', 'title': 'Dir Path', 'type': 'string'},
+        #  'pattern': {'description': 'Unix shell regex, where * matches everything.', 'title': 'Pattern', 'type': 'string'}}
+        @property
+        def args(self) -> dict[str, dict[str, str]]: ...
```
This is an example with 2 string arguments, but if there are dict or list arguments, I imagine this could be more deeply nested and include value other than strings. Could we just make this `dict[str, Any]` to support any JSON schema?

DouweM reviewed (commented):
@matthewfranglen Thanks Matthew! In addition to the comments I've left, can you please add docs for this in `docs/tools.md`?

DouweM reviewed (changes requested):
@matthewfranglen Thanks Matthew! In addition to the comments I've left, can you please add docs for this in `docs/tools.md`?

matthewfranglen commented on tests/test_tools.py:
```diff
@@ -12,7 +12,14 @@
 
 from pydantic_ai import Agent, RunContext, Tool, ToolOutput, UserError
-from pydantic_ai.messages import ModelMessage, ModelRequest, ModelResponse, TextPart, ToolCallPart, ToolReturnPart
+from pydantic_ai.messages import (
```
Sorry I'll stop this happening, I agree it's annoying.

DouweM replied:
@matthewfranglen That's weird, in `pyproject.toml` we set `line-length` for `tool.ruff` to `120`, but in your case it seems to want 80.

This shouldn't make a difference, but can you try `make format`? Is it possible your IDE or something else is overriding `line-length`?

matthewfranglen commented on pydantic_ai_slim/pydantic_ai/tools.py:
```diff
@@ -304,6 +325,79 @@ async def prep_my_tool(
+        def proxy(*args: Any, **kwargs: Any) -> str:
+            for argument, key in zip(args, inputs.keys()):
+                kwargs[key] = argument
+            for name, default_value in defaults.items():
+                if name in kwargs:
+                    continue
+                kwargs[name] = default_value
```
That change mutates defaults:

```
a = {1: "one"}
b = {1: "two"}
c = a.update(b)
print(a)
>> {1: 'two'}
```

it can be simplified, a python 3.9 compatible approach is `{**defaults, **kwargs}`, later pythons support `defaults | kwargs`.

matthewfranglen replied:
Ooh no dictionary merging with `|` was introduced in python 3.9. Neat.

matthewfranglen commented on pydantic_ai_slim/pydantic_ai/tools.py:
```diff
@@ -9,16 +10,34 @@
+        def name(self) -> str: ...
+
+        @property
+        def description(self) -> str: ...
```
The tools in langchain extend from `langchain_core.tools.BaseTool`. `BaseTool` extends from `pydantic.BaseModel` (via `RunnableSerializable` -> `Serializable`). At the time of writing `name` and `description` are required: https://github.com/langchain-ai/langchain/blob/b808d272849516d1c0b902421edf1933a87d9afa/libs/core/langchain_core/tools/base.py#L356-L362

matthewfranglen commented on pydantic_ai_slim/pydantic_ai/tools.py:
```diff
@@ -304,6 +325,79 @@ async def prep_my_tool(
+        """Creates a Pydantic tool from a function and a JSON schema.
+
+        Args:
+            function: The function to call
```
There will be no validation, I agree. While the langchain adapter I have written will only use keyword arguments the schema of an arbitrary function may require positional arguments.

DouweM replied:
@matthewfranglen But since we're setting `FunctionSchema.positional_fields` to an empty list, won't `_call_args` only ever create a kwargs dict, meaning that a function that takes positional args won't get them?

matthewfranglen replied:
The json schema for a positional only function does not communicate the positional nature of the argument. As such I have added the clarifying statement that you have suggested.

matthewfranglen commented on tests/test_tools.py:
```diff
@@ -1048,3 +1079,252 @@ def foobar(ctx: RunContext[int], x: int, y: str) -> str:
+
+    def run(
+        self,
+        tool_input: Union[str, dict[str, Any]],
```
I've done a little exploratory testing for this. The langchain tools that take a single string are like this example one:

https://python.langchain.com/docs/how_to/custom_tools/#creating-tools-from-runnables

```python
from langchain_core.language_models import GenericFakeChatModel
from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate

prompt = ChatPromptTemplate.from_messages(
    [("human", "Hello. Please respond in the style of {answer_style}.")]
)

# Placeholder LLM
llm = GenericFakeChatModel(messages=iter(["hello matey"]))

chain = prompt | llm | StrOutputParser()

as_tool = chain.as_tool(
    name="Style responder", description="Description of when to use tool."
)
```

If I stick this into a test then the arguments of it come out as:

```json
{'answer_style': {'title': 'Answer Style', 'type': 'string'}}
```

If I convert this to a Pydantic-AI tool then the json_schema is:

```json
{'type': 'object', 'properties': {'answer_style': {'title': 'Answer Style', 'type': 'string'}}, 'additionalProperties': False, 'required': ['answer_style']}
```

(and it works).

If I create a simple function that takes a single string argument and convert it to a tool I see much the same:

```python
def fn(arg: str) -> str:
    """Say hello

    Args:
        arg (str): the argument
    """
    return "hello"

py_tool = Tool(fn)
print(py_tool.function_schema.json_schema)
```

```json
{'additionalProperties': False, 'properties': {'arg': {'description': 'the argument', 'type': 'string'}}, 'required': ['arg'], 'type': 'object'}
```

So this current code supports LangChain tools that take a single string. I noticed in the Pydantic-AI tool documentation that there is an optimization that is performed for single pydantic type as the argument, where the fields of that type are lifted:

> If a tool has a single parameter that can be represented as an object in JSON schema (e.g. dataclass, TypedDict, pydantic model), the schema for the tool is simplified to be just that object.

https://ai.pydantic.dev/tools/#function-tools-and-schema

I have not explicitly applied that optimization.

matthewfranglen replied:
I've been testing the conversion of langchain functions and I have found that defining a function that takes a pydantic object as the argument results in a schema that uses references. The current code does not handle that correctly. I have a fix for it.

However the lifting is not applied. For this function:

```python
class ComplexType(BaseModel):
    name: str
    age: int

def fn(data: ComplexType) -> str:
    """Does foobar really good"""
    return "hello"
```

The langchain -> pydantic tool has a schema of:

```json
{
  "$defs": {
    "ComplexType": {
      "properties": {
        "name": {
          "title": "Name",
          "type": "string"
        },
        "age": {
          "title": "Age",
          "type": "integer"
        }
      },
      "required": [
        "name",
        "age"
      ],
      "title": "ComplexType",
      "type": "object"
    }
  },
  "description": "Does foobar really good",
  "properties": {
    "data": {
      "$ref": "#/$defs/ComplexType"
    }
  },
  "required": [
    "data"
  ],
  "title": "fn",
  "type": "object",
  "additionalProperties": false
}
```

While the pure pydantic tool has a schema of:

```
{
  "properties": {
    "name": {
      "type": "string"
    },
    "age": {
      "type": "integer"
    }
  },
  "required": [
    "name",
    "age"
  ],
  "title": "ComplexType",
  "type": "object"
}
```

The reference approach allows for this Peano number definition:

```python
class PeanoNumber(BaseModel):
    prev: PeanoNumber | None
PeanoNumber.model_rebuild()
```

The langchain function determines that the schema of this is:

```json
{
  "$defs": {
    "PeanoNumber": {
      "properties": {
        "inner": {
          "anyOf": [
            {
              "$ref": "#/$defs/PeanoNumber"
            },
            {
              "type": "null"
            }
          ]
        }
      },
      "required": [
        "inner"
      ],
      "title": "PeanoNumber",
      "type": "object"
    }
  },
  "$ref": "#/$defs/PeanoNumber"
}
```

However getting the args representation for this fails, so the conversion to a Pydantic-AI tool would fail. Equally the use of this type directly in a Pydantic-AI Tool fails with a Name Error: `NameError: name 'PeanoNumber' is not defined`.

In conclusion the handling of single argument complex types could be improved.

Kludex commented on docs/tools.md:
```diff
@@ -624,3 +624,30 @@ def my_flaky_tool(query: str) -> str:
+
+## Use LangChain Tools {#langchain-tools}
+
+We love LangChain and think it has a very compelling suite of tools. To import a tool from LangChain, use the `from_langchain()` method.
```
I think we should mention that it's a good way to transition to PydanticAI.

matthewfranglen replied:
If you want to propose alternative text then please do. Since the documentation is presenting a specific view of the pydantic product that you want to preserve I think it would be faster for all involved if any documentation changes were provided verbatim.

matthewfranglen commented on pydantic_ai_slim/pydantic_ai/tools.py:
```diff
@@ -304,6 +327,78 @@ async def prep_my_tool(
+        def proxy(*args: Any, **kwargs: Any) -> str:
+            assert not args, 'This should always be called with kwargs'
+            kwargs = defaults | kwargs
+            return langchain_tool.run(kwargs)
```
I tried writing a tool that immediately throws an exception and invoking it using a langchain agent. The agent invocation appeared to hang. Maybe this is something that could be further investigated in future.

DouweM replied:
OK, just wanted to make sure there are no obvious exceptions for us to catch and wrap and reraise here.

matthewfranglen commented on docs/tools.md:
````diff
@@ -624,3 +624,30 @@ def my_flaky_tool(query: str) -> str:
 ```
 Raising `ModelRetry` also generates a `RetryPromptPart` containing the exception message, which is sent back to the LLM to guide its next attempt. Both `ValidationError` and `ModelRetry` respect the `retries` setting configured on the `Tool` or `Agent`.
+
+## Use LangChain Tools {#langchain-tools}
````
I've added it, with an example bit of code. The linter for the docs is very unhelpful and I can't see how to fix it:

```
ruff failed:
  docs/tools.md:415:1: I001 [*] Import block is un-sorted or un-formatted
    |
  1 | from pydantic_ai.tools import Tool
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ I001
  2 |
  3 | def foobar(**kwargs) -> str:
```

DouweM replied:
That's weird, there's just one import so nothing to sort :/ You can leave this as is for now and I see how to fix it before I merge.

DouweM commented on docs/tools.md:
```diff
@@ -409,6 +409,33 @@ print(test_model.last_model_request_parameters.function_tools)
+        'required': ['a', 'b'],
+        'type': 'object',
+    }
+)
```
Can you please add an example of passing this to an agent, just so people don't get lost on what to do wit this new `tool`?

DouweM commented on pydantic_ai_slim/pydantic_ai/tools.py:
```diff
@@ -304,6 +327,80 @@ async def prep_my_tool(
+        )
+
+    @classmethod
+    def from_langchain(cls, langchain_tool: LangChainTool) -> Self:
```
What do you think about moving this off of `Tool` and into a new module like `pydantic_ai.ext.langchain`? We will likely have more wrappers and convenience methods for other frameworks in the future, so my colleague @Viicos correctly pointed out that we should probably start moving them into their own namespace already instead of cluttering up the regular public classes.

This wouldn't need to be class method in that case, I imagine:

```py
from pydantic_ai import Agent
from pydantic_ai.ext.langchain import langchain_tool
from langchain_community.tools import DuckDuckGoSearchRun

agent = Agent(..., tools=[langchain_tool(DuckDuckGoSearchRun])
```
